docs/source/genetic/config.rst
Configuration
--------------

The genetic algorithm (GA) has a helper class that is used to parse `.yml` files to read the configuration that has to be used.

This configuration :class:`GeneticBaseConfig<mango.models.genetic.config.GeneticBaseConfig>` class can be extended and modified at will.

The parameters that can be modified are divided by sections.

Main section
============

In this section we have the main controls of the genetic algorithm:

- **population_size**: The size of the population that will be used in the GA. It has to be an integer value and it defaults to 100.
- **max_generations**: The maximum number of generations that will be used in the GA. It has to be an integer value and it defaults to 500.
- **optimization_objective**: The objective of the optimization. It can be `max` or `min`. It has to be a string and can only be either `max` or `min`.
- **selection**: The selection method that will be used in the GA. It can be any of the following values: `random`, `elitism`, `rank`, `order`, `roulette` or `tournament`.
- **crossover**: The crossover method that will be used in the GA.It can be any of the following values: `one-split`, `two-split`, `mask`, `linear`, `flat`, `blend` or `gaussian`.
- **replacement**: The replacement method that will be used in the GA. It can be any of the following values: `random`, `only-offspring`, `elitist` or `elitist-stochastic`.
- **mutation_control**: The mutation control method that will be used in the GA. It can be `static`, `gene-based`, `population-based` or `adaptative`.
- **mutation_base_rate**: The base mutation rate that will be used in the GA. It can be any float number between 0 and 1 and it defaults to 0.1.

An example of the `.yml` file can be seen following:

.. code-block:: yaml

    [main]
    population_size = 100
    max_generations = 100
    optimization_objective = max
    selection = tournament
    crossover = one_split
    replacement = elitism
    mutation_control = adaptative
    mutation_base_rate = 0.1


Individual section
==================

The section for the individual has the following parameters:

- **encoding**: The encoding that will be used in the GA. It can be `binary`, `integer` or `real`.
- **gene_length**: The length of the genome that will be used in the GA. It has to be an integer value and it defaults to 0.
- **gene_min_value**: The minimum value that a gene can have. It has to be a float value.
- **gene_max_value**: The maximum value that a gene can have. It has to be a float value.

An example of the `.yml` file can be seen following:

.. code-block:: yaml

    [individual]
    encoding = binary
    gene_length = 10
    gene_min_value = 0.0
    gene_max_value = 1.0


Selection section
=================

The section for the selection has the following parameters:

- **elitism_size**: The size of the elitism that will be used in the GA. It has to be an integer value and it defaults to 20.
- **tournament_size**: The size of the tournament that will be used in the GA. It has to be an integer value and it defaults to 2.
- **rank_pressure**: The rank pressure that will be used in the GA. It has to be a float value between 1 and 2 and defaults to 2.

An example of the `.yml` file can be seen following:

.. code-block:: yaml

    [selection]
    elitism_size = 20
    tournament_size = 2
    rank_pressure = 2.0

Crossover section
=================

The section for the crossover has the following parameters:

- **offspring_size**: The size of the offspring that will be used in the GA. It has to be an integer value and it defaults to 100.
- **blend_expansion**: The blend expansion that will be used in the GA. It has to be a float value and defaults to 0.5.

An example of the `.yml` file can be seen following:

.. code-block:: yaml

    [crossover]
    offspring_size = 100
    blend_expansion = 0.5

Mutation section
================

The section for the mutation has the following parameters:

- **generation_adaptative**: The generation adaptative that will be used in the GA. It has to be an integer value and it defaults to 10.

An example of the `.yml` file can be seen following:

.. code-block:: yaml

    [mutation]
    generation_adaptative = 10


changelog.rst
Version 0.2.1
==============

**Released on Jan 8, 2024.**

Small bugfixes on genetic algorithms and code reordering for documentation.


Version 0.2.0
==============

**Released on Dec 21, 2023.**

This version includes a new module for genetic algorithms and more!

**New features:**

- Added genetic algorithms that can be used in a wide range of problems.
- Added a new module to perform SHAP analysis.
- Added a function to pivot the calendar.
- Added correlation calculation between series.

**Bugfixes:**

- Minor bugfixes.

Version 0.1.3
==============

**Released on Nov 17, 2023.**

This version includes a small bugfix.

**Bugfixes:**

- Fixed a bug on save button with unique keys on the dashboard.

Version 0.1.2
==============

**Released on Nov 17, 2023.**

This version includes a function to calculate distances with the haversine formula.

**New features:**

- Added a function to calculate distances with the haversine formula.
- Added the possibility to edit excels, csv and json files in the streamlit app.

**Bugfixes:**

- Fixed a bug on streamlit app saving the config file.

Version 0.1.1
==============

**Released on Nov 8, 2023.**

This version includes a small bugfix.

**Bugfixes:**

- Fixed a bug on left join on class Table.
- Fixed a small bug on AEMET pydantic validation.

Version 0.1.0
==============

**Released on Nov 7, 2023.**

This version includes new modules and bugfixes.

**New features:**

- Added improvements to ``Table`` class.
- Adapted ``load_excel`` function to return a list of dictionaries.
- Added new date functions.
- Added a REST client to the AEMET to gather meteorological data.
- Added pydantic validation as a decorator that can be used in any function.
- Added calendar datasets for Spain.
- Moved from setup.py to pyproject.toml.

**Bugfixes:**

- Fixed a bug on left join on class Table.

**Breaking changes:**

- Removed pandas as a main dependency of the library. It is still needed in some submodules.
- Drop images module.

Version 0.0.6
==============

**Released on May 31, 2023.**

Improvements to Table class

Version 0.0.5
==============

**Released on Apr 26, 2023.**

Fixed some bugs on ``Table`` class.

**Bugfixes:**

- Fixed error on ``mutate`` when table had one row.
- Allow group by None in ``sum_all`` and ``summarise``.
- ``as_list`` applied on a dict returns a list of the dict instead of the keys.

Version 0.0.4
==============

**Released on Apr 19, 2023.**

This versions includes changes to the logger class and methods.

**New features:**

- ``Chrono`` class can receive a logger name to be used.
- ``log_time`` decorator can receive a logger name to be used.

Version 0.0.3
==============

**Released on Apr 18, 2023.**

Minor changes to logging class ``Chrono`` and to the default logger.

**New features:**

- Minor changes to logging class ``Chrono``: ``stop`` now reports the duration.
- Default logger has a info level set up.

Version 0.0.2
==============

**Released on Apr 11, 2023.**

This version includes a new class ``Table`` and a new module to perform requests to ArcGIS.

**New features:**

- Added class ``Table``.
- Added direct requests for arcgis od matrix calculation.

Version 0.0.1
==============

Released on Mar 9, 2023.

This is the first version of mango library.


mango/models/genetic/config/config.py
from mango.config import BaseConfig, ConfigParameter
from mango.models.genetic.const import (
    INDIVIDUAL_ENCODINGS,
    SELECTION_METHODS,
    CROSSOVER_METHODS,
    REPLACEMENT_METHODS,
    MUTATION_CONTROL_METHODS,
)


class GeneticBaseConfig(BaseConfig):
    __params = {
        "main": [
            ConfigParameter("population_size", int, 100),
            ConfigParameter("max_generations", int, 500),
            ConfigParameter("optimization_objective", str, validate=["max", "min"]),
            ConfigParameter("selection", str, validate=SELECTION_METHODS),
            ConfigParameter("crossover", str, validate=CROSSOVER_METHODS),
            ConfigParameter("replacement", str, validate=REPLACEMENT_METHODS),
            ConfigParameter(
                "mutation_control",
                str,
                validate=MUTATION_CONTROL_METHODS,
                default="none",
            ),
            ConfigParameter(
                "mutation_base_rate",
                float,
                0.1,
                required=False,
                min_value=0,
                max_value=1,
            ),
            ConfigParameter(
                "checkpoint_save_path",
                str,
                None,
                required=False,
            ),
            ConfigParameter(
                "checkpoint_save_period",
                int,
                None,
                required=False,
                min_value=0,
            ),
        ],
        "individual": [
            ConfigParameter("encoding", str, validate=INDIVIDUAL_ENCODINGS),
            ConfigParameter("gene_length", int, default=0),
            ConfigParameter("gene_min_value", float),
            ConfigParameter("gene_max_value", float),
        ],
        "selection": [
            ConfigParameter("elitism_size", int, 20, required=False),
            ConfigParameter("tournament_size", int, 2, required=False),
            ConfigParameter(
                "rank_pressure",
                float,
                2.0,
                required=False,
                min_value=1.0,
                max_value=2.0,
            ),
        ],
        "crossover": [
            ConfigParameter("offspring_size", int, 100, required=False),
            ConfigParameter("blend_expansion", float, 0.1, required=False),
        ],
        "mutation": [
            ConfigParameter("generation_adaptative", int, default=10, required=False),
        ],
    }

    def __init__(self, filename):
        super().__init__(filename, self.__params)
        # Add validation for checkpoint_save_path and checkpoint_save_rate
        if (self("checkpoint_save_path") is None) != (
            self("checkpoint_save_period") is None
        ):
            raise ValueError(
                "Both 'checkpoint_save_path' and 'checkpoint_save_period' must be set together."
            )
